werkzeug/test.py

```diff
@@ -338,6 +340,27 @@ def __init__(self, path='/', base_url=None, query_string=None,
                         self._add_file_from_data(key, value)
                     else:
                         self.form.setlistdefault(key).append(value)
+        if json:
+            input_count = 0
```

Comment:
This logic should be moved before all other checks, it also makes the equivalent check in the `if data` block obsolete.

Also it can be reduced to `if len(list(filter(bool, (data, json, input_stream)))) != 1`.